File: C/LeetCode/002-AddTwoNumbers.c
#include "002-AddTwoNumbers.h"

int getValueAndMoveNode(struct ListNode **node)
{
	if (*node == NULL) { return 0; }
	int value = (*node)->val;
	*node = (*node)->next;
	return value;
}

struct ListNode *calculateNext(struct ListNode **node, struct ListNode **l1, struct ListNode **l2)
{
	int value1 = getValueAndMoveNode(l1);
	int value2 = getValueAndMoveNode(l2);

	int sum = value1 + value2 + (*node)->val;

	if (sum >= 10)
	{
		(*node)->val = sum % 10;
		(*node)->next = (struct ListNode *)malloc(sizeof(struct ListNode));
		(*node)->next->val = 1;
		(*node)->next->next = NULL;
	}
	else
	{
		(*node)->val = sum;
		(*node)->next = (struct ListNode *)malloc(sizeof(struct ListNode));
		(*node)->next->val = 0;
		(*node)->next->next = NULL;
	}

	return *node;
}

struct ListNode *addTwoNumbers(struct ListNode *l1, struct ListNode *l2)
{
	struct ListNode *first, *current;
	first = (struct ListNode *)malloc(sizeof(struct ListNode));
	first->val = -1;
	first->next = (struct ListNode *)malloc(sizeof(struct ListNode));
	first->next->val = 0;
	first->next->next = NULL;
	current = first;

	while (l1 != NULL || l2 != NULL)
	{
		current = calculateNext(&(current->next), &l1, &l2);
	}

	if (current->next->val == 0)
	{
		struct ListNode *temp = current->next;
		current->next = NULL;
		free(temp);
	}
	return first->next;
}

File: CSharp/LeetCode/002-AddTwoNumbers.cs
﻿namespace LeetCode
{
    public class _002_AddTwoNumbers
    {
        public ListNode AddTwoNumbers(ListNode l1, ListNode l2)
        {
            var first = new ListNode(-1);
            first.next = new ListNode(0);
            var current = first;

            while (l1 != null || l2 != null)
            {
                var value1 = l1 == null ? 0 : l1.val;
                var value2 = l2 == null ? 0 : l2.val;

                current = current.next;
                var sum = value1 + value2 + current.val;

                if (sum >= 10)
                {
                    current.val = sum % 10;
                    current.next = new ListNode(1);
                }
                else
                {
                    current.val = sum;
                    current.next = new ListNode(0);
                }

                l1 = l1 == null ? null : l1.next;
                l2 = l2 == null ? null : l2.next;
            }

            if (current.next.val == 0)
            {
                current.next = null;
            }

            return first.next;
        }
    }
}
